store/stat_store.go

```diff
@@ -0,0 +1,442 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"container/list"
+	"fmt"
+	"math"
+	"sort"
+	"sync"
+	"time"
+)
+
+func NewStatStore(epsilon uint64, resolution time.Duration, windowDuration uint8, supportedPercentiles []float64) *statStore {
+	return &statStore{
+		buffer:               list.New(),
+		epsilon:              epsilon,
+		resolution:           resolution,
+		windowDuration:       windowDuration,
+		supportedPercentiles: supportedPercentiles,
+	}
+}
+
+// statStore is an in-memory rolling timeseries buffer that allows the extraction of -
+// segments of the stored timeseries and derived stats.
+// The tradeoff between space and precision can be configured through epsilon and resolution.
+// statStore only retains information about the latest TimePoints that are within its
+// specified duration.
+// statStore is a TimeStore-like object that does not implement the TimeStore interface.
+type statStore struct {
+	// start is the start of the time window.
+	start time.Time
+
+	// buffer is a list of tpBucket that is sequenced in a time-descending order, meaning that
+	// Front points to the latest tpBucket and Back to the oldest one.
+	buffer *list.List
+
+	// A RWMutex guards all operations on the underlying buffer.
+	sync.RWMutex
+
+	// epsilon is the acceptable error difference for the storage of TimePoints.
+	// Increasing epsilon decreases the memory usage of the statStore, at the cost of precision.
+	// The precision of max is not affected.
+	epsilon uint64
```

Comment:
epsilon represents error margin in value. Since all TimePoint values are assumed to be uint64, then having a floating point epsilon makes no sense. For example, setting epsilon to 50.53412 would effectively be the same as setting it to 50. Also, epsilon cannot be a percentage offset as that would cause bucket overlaps. For example, a bucket with average value of 100 and epsilon of 0.1 would overlap with a bucket with value of 110.